casper/src/main/scala/coop/rchain/casper/util/rholang/SystemDeployUtil.scala

```diff
@@ -0,0 +1,47 @@
+package coop.rchain.casper.util.rholang
+
+import com.google.protobuf.ByteString
+
+import coop.rchain.casper.genesis.contracts.Validator
+import coop.rchain.casper.protocol.DeployData
+import coop.rchain.crypto.hash.Blake2b512Random
+import coop.rchain.crypto.signatures.Signed
+import coop.rchain.crypto.PublicKey
+
+object SystemDeployUtil {
+  import coop.rchain.models.Validator.Validator
+  // Currently we have 4 system deploys -> refund, preCharge, closeBlock, Slashing
+  // In every user deploy, the rnode would do the preCharge first, then execute the
+  // user deploy and do the refund at last.
+  //
+  // The refund and preCharge system deploy
+  // would use user deploy signature to generate the system deploy. The random seed of
+  // the refund and preCharge has to be exactly the same to make sure replay the user
+  // deploy would come out the exact same result.
+  //
+  // As for closeBlock and slashing, the rnode would execute closeBlock system deploy in every block.
+  // The closeBlock system deploy would use the sender and the seqNum of the validator.
+
+  def generateSystemDeployRandomSeed(sender: Validator, seqNum: Int): Blake2b512Random =
+    Tools.rng(sender.toByteArray() ++ Array[Byte](seqNum.toByte))
```

Comment:
`Array[Byte](seqNum.toByte)` will truncate prefix with this conversion. I'm thinking that we can use protobuf to create fixed length serialization to bytes.
Something like this should work.
```scala
import java.io.{ByteArrayInputStream, ByteArrayOutputStream}

import com.google.protobuf.{CodedInputStream, CodedOutputStream}

def serializeInt64Fixed(value: Long): Array[Byte] = {
  val stream = new ByteArrayOutputStream
  val proto = CodedOutputStream.newInstance(stream)
  proto.writeFixed64NoTag(value)
  proto.flush()
  stream.toByteArray
}

def deserializeInt64Fixed(bytes: Array[Byte]): Long = {
  val stream = new ByteArrayInputStream(bytes)
  val proto = CodedInputStream.newInstance(stream)
  proto.readSFixed64
}

val values = List(1234567890123456789L, 1L)

val valuesSer = values.map(serializeInt64Fixed)
// List(Array(21, -127, -23, 125, -12, 16, 34, 17), Array(1, 0, 0, 0, 0, 0, 0, 0))

val valuesCopy = valuesSer.map(deserializeInt64Fixed)
// List(1234567890123456789, 1)

assert(values == valuesCopy)
```

Reply:
My concerns are similar as with [ETH signatures, hashes and byte messages](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md#signatures-and-hashing-overview).
We need to be careful of which parts the "message" (seed) is constructed. With variable length of messages, length can be included to prevent collisions. We don't need that but we need to take care about the length of each part of the message and total size. This is why I'm suggesting protobuf fixed serializer.

So for a block seed it would be enough to have: `PREFIX ++ PublicKey ++ seqNum` serialized with protobuf. This way we can be completely sure that collision cannot happen. And we can you the same technique if we need seed for other purposes in the future.

Reply:
Are you sure the bytes would be the same? It seems it is different.  Replay would be wrong in this way.

Reply:
> My concerns are similar as with [ETH signatures, hashes and byte messages](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md#signatures-and-hashing-overview).
> We need to be careful of which parts the "message" (seed) is constructed. With variable length of messages, length can be included to prevent collisions. We don't need that but we need to take care about the length of each part of the message and total size. This is why I'm suggesting protobuf fixed serializer.
> 
> So for a block seed it would be enough to have: `PREFIX ++ PublicKey ++ seqNum` serialized with protobuf. This way we can be completely sure that collision cannot happen. And we can you the same technique if we need seed for other purposes in the future.

Your point totally makes sense and good. I was thinking to use `com.google.protobuf.wrappers.Int32Value`.But I am not sure if it is fixed length

Reply:
With `Int32Value` wrapper you still need to write to OutputStream.
https://github.com/yeriomin/play-store-api/blob/4ec5513/src/main/java/com/google/protobuf/Int64Value.java#L56-L61
I find it easier to use `CodedOutputStream` and `CodedInputStream` directly without the wrapper where any supported type can have its own method.